### PR TITLE
Native wrapper

### PR DESCRIPTION
**Pull request by modmuss50**
This is take 2 of a native wrapper, now in nim. This should be more approachable than c++, less code and more portable. What is in this PR should be mostly fully feature complete (other than x86 + testing)

See #60 for my first proof of concept.

### CODE REVIEW

**Comment by liach** [comment]
As long as this is built by ci properly this should be good

**Comment by modmuss50** [comment]
Yes, my plan is to built this from github actions running on windows.

File: wrapper.nim [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    # Forward the exit code
+    let exitCode = process.peekExitCode()
+    echo &"Fabric installer exit code: {exitCode}"
+    system.quit(exitCode)
+
+when isMainModule:
+    echo "Fabric installer bootstrap"
+
+    when not defined windows:
+        echo "Only windows is currenly supported"
```
Comment (i509VCB) [inline wrapper.nim:73]:
```suggestion
        echo "Only windows is currently supported"
```

File: wrapper.nim [hunk 2]
```diff
@@ -0,0 +1,85 @@
+import os
+import osproc
+import streams
+import strformat
+import dialogs
+
+# Statically includes the installer binary
+const installerBinary = staticRead("/build/libs/fabric-installer-0.7.0.jar")
```
Comment (i509VCB) [inline wrapper.nim:8]: staticRead links to the jar or copies it raw into the program?

**Comment by liach** [comment]
Oh right this is just an exe wrapper. I thought this might have some dmg wrapper for mac :laughing: all great

File: .gitignore [hunk 3]
```diff
@@ -1,23 +1,23 @@
...
+!/gradle.properties
+!/settings.gradle
+!/gradlew
+!/gradlew.bat
+!/HEADER
+!/Jenkinsfile
+!/l4j-manifest.xml
+!/LICENSE
+!/README.md
+!/wrapper.nim
```
Comment (i509VCB) [inline .gitignore:22]: I would put a comment in the gitignore signifying native source files vs java source files.

File: wrapper.nim (hunk 2 above)
Comment (i509VCB) [reply wrapper.nim:8]: Also the value in the string should probably get set at compile time.

Comment (modmuss50) [reply wrapper.nim:8]: It copies it into the exe, and yes, ill prob get gradle to spit out a jar without the version.

**Comment by modmuss50** [comment]
@liach there is not reason this cannot be expanded to support mac and linux, the current plan to is to solve the issues around launch4j and windows.

**Comment by comp500** [comment]
It'd probably be a good idea to add a link to download Java in the error message, or the wiki tutorial for doing so.

**Comment by liach** [comment]
Will this move to the native bootstrap repo?

**Comment by modmuss50** [comment]
Yes, there will be some work to do here as well. Thanks to a tip I found that jars and exe's can just be botled togeather into one and it works just fine. So there wont be a need to rebuild the native stuff each time for the installer.

**Comment by modmuss50** [comment]
Closed, see my cleanup PR for the tiny gradle change needed for this.

**Pull request closed without merging**